File: 백준/Gold/11967. 불켜기/불켜기.swift
import Foundation

let input = readLine()!.split(separator: " ").map { Int($0)! }
let n = input[0], m = input[1]


var switchMap = Array(repeating: Array(repeating: [(Int, Int)](), count: n + 1), count: n + 1)
var isOn = Array(repeating: Array(repeating: false, count: n + 1), count: n + 1)
var visited = Array(repeating: Array(repeating: false, count: n + 1), count: n + 1)

let dx = [-1, 1, 0, 0]
let dy = [0, 0, -1, 1]

for _ in 0..<m {
    let line = readLine()!.split(separator: " ").map { Int($0)! }
    let (x, y, a, b) = (line[0], line[1], line[2], line[3])
    switchMap[x][y].append((a, b))
}
isOn[1][1] = true
visited[1][1] = true
var queue = [(1, 1)]
var idx = 0

while idx < queue.count {
    let (x, y) = queue[idx]
    idx += 1

    
    for (a, b) in switchMap[x][y] {
        if !isOn[a][b] {
            isOn[a][b] = true
            for i in 0..<4 {
                let nx = a + dx[i]
                let ny = b + dy[i]
                if 1 <= nx, nx <= n, 1 <= ny, ny <= n {
                    if visited[nx][ny] {
                        visited[a][b] = true
                        queue.append((a, b))
                        break
                    }
                }
            }
        }
    }

    
    for i in 0..<4 {
        let nx = x + dx[i]
        let ny = y + dy[i]

        if 1 <= nx, nx <= n, 1 <= ny, ny <= n {
            if isOn[nx][ny] && !visited[nx][ny] {
                visited[nx][ny] = true
                queue.append((nx, ny))
            }
        }
    }
}

// 불켜진 방세기
var result = 0
for i in 1...n {
    for j in 1...n {
        if isOn[i][j] {
            result += 1
        }
    }
}
print(result)

File: 백준/Gold/11967. 불켜기/불켜기.py
import sys
from collections import deque
input=sys.stdin.readline

N,M=map(int,input().split())

graph=[[[] for _ in range(N)] for _ in range(N)]


for _ in range(M):
    x,y,a,b=map(int,input().split())
    graph[x-1][y-1].append((a-1,b-1))

light=[[False]*N for _ in range(N)]
light[0][0]=True

visited=[[False]*N for _ in range(N)]

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

def check(x, y):
    for i in range(4):
        nx, ny = x + dx[i], y + dy[i]
        if 0 <= nx < N and 0 <= ny < N and visited[nx][ny]:
            return True
    return False

def bfs():
    
    q=deque()
    q.append((0,0))
    visited[0][0]=True

    while True:
        changed=False

        while q:
            x,y=q.popleft()

            for a,b in graph[x][y]:
                if not light[a][b]:
                    light[a][b]=True
                    if check(a,b):
                        visited[a][b] = True
                        q.append((a,b))
                        changed=True

            for i in range(4):
                nx,ny=x+dx[i],y+dy[i]
                if 0<=nx<N and 0<=ny<N:
                    if light[nx][ny] and not visited[nx][ny]:
                        visited[nx][ny]=True
                        q.append((nx,ny))
                        changed=True
        if not changed:
            break
bfs()
cnt=0
for row in light:
    cnt+=row.count(True)
print(cnt)